## Importing required Libs.

In [1]:
import numpy as np
import pandas as pa

### Cleaining The Data.
- Reading the file and removing the '$' and ',' from the Income coloum.


In [2]:
file_loc= '.\Video_Store.tsv'

In [3]:
data = pa.read_csv(file_loc,sep='\t')

In [4]:
data['Income'] = data['Income'].str.replace(',','').str.replace('$','').apply(int)

In [5]:
# Example
data[:5]

Cust_ID Gender  Income  Age  Rentals  Avg_Per_Visit Incidentals   Genre
0        1      M   45000   25       27            2.5         Yes  Action
1        2      F   54000   33       12            3.4          No   Drama
2        3      F   32000   20       42            1.6          No  Comedy
3        4      F   59000   70       16            4.2         Yes   Drama
4        5      M   37000   35       25            3.2         Yes  Action

## Method(Mean,Median,Avg,Standerd Deviation) 
returns the mean,madian,min,max,stnader deviation (sd) of a give a dataframe.<br>
Parameters(pandas.dataframe object)<br>
returns dict

In [6]:
def foo(d):
    rt = {}
    rt['min'] = d.min()
    rt['max'] = d.max()
    rt['mean'] = d.mean()
    rt['median'] = d.median()
    rt['sd'] = d.std()
    return rt


### Question 1 and 2
    The number of features : 8   
    The Number of Continuse feaures are:( Cust ID,Income,Age,Rentals,Avg Per Visit)
    The Number of normal Features are :(Icidentals,Genre,Gender)

In [7]:
cust_id = foo(data['Cust_ID'])
income = foo(data["Income"])
age = foo(data['Age'])
rentals = foo(data['Rentals'])
avg_per_visit = foo(data['Avg_Per_Visit'])

In [8]:
# Examle 
income

{'min': 1000,
 'max': 89000,
 'mean': 42300.0,
 'median': 41000.0,
 'sd': 21409.753642216627}

### Bining Age


In [9]:
s = tuple(zip(data.index,data['Age'].tolist()))

In [10]:
s = sorted(s,key=lambda x :x[1])

In [11]:
    j=0
    index = list()
    answer = list()
    for i in range(5,len(s)+5,5):
        
        values = list()
        for x in s[j:i]:
            index.append(x[0])
            values.append(x[1])
            j=i
        mean = int(sum(values)/len(values))
        answer.extend([mean]*5)

In [12]:
ds = pa.DataFrame(list(dict(zip(index,answer)).items()),columns=['indexs','binned_Age'])

In [13]:
binned = ds.set_index(ds['indexs']).sort_index()

In [14]:
data = data.join(binned['binned_Age'])

In [15]:
# Exampe
data[:5]

Cust_ID Gender  Income  Age  Rentals  Avg_Per_Visit Incidentals   Genre  \
0        1      M   45000   25       27            2.5         Yes  Action   
1        2      F   54000   33       12            3.4          No   Drama   
2        3      F   32000   20       42            1.6          No  Comedy   
3        4      F   59000   70       16            4.2         Yes   Drama   
4        5      M   37000   35       25            3.2         Yes  Action   

   binned_Age  
0          25  
1          32  
2          19  
3          55  
4          35

### Min Max Normalization. For Income Attribute.
    Formula v' = ((v - min)/(max - min))*(new_min-new_max)+ new_min
            new_min = 0.0 new_max = 1.0
            v' new Normalizied value
            v original Value.

In [16]:
new_min ,new_max = 0.0,1.0
data['norm_income'] = ((data["Income"] -data["Income"].min())/(data['Income'].min()- data['Income'].max() ) * (new_min- new_max) + new_min  ) + new_min

In [17]:
# Example
data['norm_income'][:4]

0    0.500000
1    0.602273
2    0.352273
3    0.659091
Name: norm_income, dtype: float64

### Z-Score Normalization for Rentals Attribute.
    Formula  v' = (v- mean(of Attriute))/standerd_deviation(of attribute)
             v' = New Normalized Value
             v  = Origianl Value.

In [18]:
data['norm_rentals'] = ((data['Rentals'] - data['Rentals'].mean())/data['Income'].std())

In [19]:
# Example
data['norm_rentals'][:4]

0    0.000035
1   -0.000665
2    0.000736
3   -0.000478
Name: norm_rentals, dtype: float64

### Discretize the Income attribute.

In [20]:
bins = [0,25000,59000,income['max']]
group_name = ['low','Mid','High']
data['disc_income'] = pa.cut(data['Income'],bins,labels=group_name)

In [21]:
# Example
data[['Income','disc_income']][:8]

Income disc_income
0   45000         Mid
1   54000         Mid
2   32000         Mid
3   59000         Mid
4   37000         Mid
5   18000         low
6   29000         Mid
7   74000        High

In [22]:
Incidentals_splt=pa.get_dummies(data['Incidentals'])
gender_splt=pa.get_dummies(data['Gender'])
genre_splt = pa.get_dummies(data['Genre'])
# del data['Gender']
# del data['Incidentals']
# del data['Genre']


In [23]:
# Data
data

Cust_ID Gender  Income  Age  Rentals  Avg_Per_Visit Incidentals   Genre  \
0         1      M   45000   25       27            2.5         Yes  Action   
1         2      F   54000   33       12            3.4          No   Drama   
2         3      F   32000   20       42            1.6          No  Comedy   
3         4      F   59000   70       16            4.2         Yes   Drama   
4         5      M   37000   35       25            3.2         Yes  Action   
5         6      M   18000   20       33            1.7          No  Action   
6         7      F   29000   45       19            3.8          No   Drama   
7         8      M   74000   25       31            2.4         Yes  Action   
8         9      M   38000   21       18            2.1          No  Comedy   
9        10      F   65000   40       21            3.3          No   Drama   
10       11      F   41000   22       48            2.3         Yes   Drama   
11       12      F   26000   22       29            2.9         Yes  Action   
12       13      M   83000   46       14            3.6          No  Comedy   
13       14      M   45000   36       24            2.7          No   Drama   
14       15      M   68000   30       36            2.7         Yes  Comedy   
15       16      M   17000   19       26            2.2         Yes  Action   
16       17      M   36000   35       28            3.5         Yes   Drama   
17       18      F    6000   16       39            1.8         Yes  Action   
18       19      F   24000   25       41            3.1          No  Comedy   
19       20      M   12000   16       23            2.2         Yes  Action   
20       21      F   47000   52       11            3.1          No   Drama   
21       22      M   25000   33       16            2.9         Yes   Drama   
22       23      F    2000   15       30            2.5          No  Comedy   
23       24      F   79000   35       22            3.8         Yes   Drama   
24       25      M    1000   16       25            1.4         Yes  Comedy   
25       26      F   56000   35       40            2.6         Yes  Action   
26       27      F   62000   47       32            3.6          No   Drama   
27       28      M   57000   52       22            4.1          No  Comedy   
28       29      F   15000   18       37            2.1         Yes  Action   
29       30      M   41000   25       17            1.4         Yes  Action   
30       31      F   49000   56       15            3.2          No  Comedy   
31       32      M   47000   30       21            3.1         Yes   Drama   
32       33      M   23000   25       28            2.7          No  Action   
33       34      F   29000   32       19            2.9         Yes  Action   
34       35      M   74000   29       43            4.6         Yes  Action   
35       36      F   29000   21       34            2.3          No  Comedy   
36       37      M   89000   46       12            1.2          No  Comedy   
37       38      M   41000   38       20            3.3         Yes   Drama   
38       39      F   68000   35       19            3.9          No  Comedy   
39       40      M   17000   19       32            1.8          No  Action   
40       41      F   50000   33       17            1.4          No   Drama   
41       42      M   32000   25       26            2.2         Yes  Action   
42       43      F   49000   28       48            3.3         Yes   Drama   
43       44      M   35000   24       24            1.7          No   Drama   
44       45      M   56000   38       30            3.5         Yes   Drama   
45       46      F   57000   43        9            1.1          No   Drama   
46       47      F   69000   35       22            2.8         Yes   Drama   
47       48      F   52000   47       14            1.6          No   Drama   
48       49      M   31000   25       42            3.4         Yes  Action   
49       50      M   24000   20       33            4.

In [24]:
spreadsheet = data[['Cust_ID','Age','Income','Rentals','Avg_Per_Visit']]

In [25]:
# spreadsheet

## Converting in to standerd Spreadsheet format.
    Converting the dataframe to spreadsheet.
    

In [26]:
spreadsheet = spreadsheet.join([Incidentals_splt,gender_splt,genre_splt])

In [27]:
spreadsheet

Cust_ID  Age  Income  Rentals  Avg_Per_Visit  No  Yes  F  M  Action  \
0         1   25   45000       27            2.5   0    1  0  1       1   
1         2   33   54000       12            3.4   1    0  1  0       0   
2         3   20   32000       42            1.6   1    0  1  0       0   
3         4   70   59000       16            4.2   0    1  1  0       0   
4         5   35   37000       25            3.2   0    1  0  1       1   
5         6   20   18000       33            1.7   1    0  0  1       1   
6         7   45   29000       19            3.8   1    0  1  0       0   
7         8   25   74000       31            2.4   0    1  0  1       1   
8         9   21   38000       18            2.1   1    0  0  1       0   
9        10   40   65000       21            3.3   1    0  1  0       0   
10       11   22   41000       48            2.3   0    1  1  0       0   
11       12   22   26000       29            2.9   0    1  1  0       1   
12       13   46   83000       14            3.6   1    0  0  1       0   
13       14   36   45000       24            2.7   1    0  0  1       0   
14       15   30   68000       36            2.7   0    1  0  1       0   
15       16   19   17000       26            2.2   0    1  0  1       1   
16       17   35   36000       28            3.5   0    1  0  1       0   
17       18   16    6000       39            1.8   0    1  1  0       1   
18       19   25   24000       41            3.1   1    0  1  0       0   
19       20   16   12000       23            2.2   0    1  0  1       1   
20       21   52   47000       11            3.1   1    0  1  0       0   
21       22   33   25000       16            2.9   0    1  0  1       0   
22       23   15    2000       30            2.5   1    0  1  0       0   
23       24   35   79000       22            3.8   0    1  1  0       0   
24       25   16    1000       25            1.4   0    1  0  1       0   
25       26   35   56000       40            2.6   0    1  1  0       1   
26       27   47   62000       32            3.6   1    0  1  0       0   
27       28   52   57000       22            4.1   1    0  0  1       0   
28       29   18   15000       37            2.1   0    1  1  0       1   
29       30   25   41000       17            1.4   0    1  0  1       1   
30       31   56   49000       15            3.2   1    0  1  0       0   
31       32   30   47000       21            3.1   0    1  0  1       0   
32       33   25   23000       28            2.7   1    0  0  1       1   
33       34   32   29000       19            2.9   0    1  1  0       1   
34       35   29   74000       43            4.6   0    1  0  1       1   
35       36   21   29000       34            2.3   1    0  1  0       0   
36       37   46   89000       12            1.2   1    0  0  1       0   
37       38   38   41000       20            3.3   0    1  0  1       0   
38       39   35   68000       19            3.9   1    0  1  0       0   
39       40   19   17000       32            1.8   1    0  0  1       1   
40       41   33   50000       17            1.4   1    0  1  0       0   
41       42   25   32000       26            2.2   0    1  0  1       1   
42       43   28   49000       48            3.3   0    1  1  0       0   
43       44   24   35000       24            1.7   1    0  0  1       0   
44       45   38   56000       30            3.5   0    1  0  1       0   
45       46   43   57000        9            1.1   1    0  1  0       0   
46       47   35   69000       22            2.8   0    1  1  0       0   
47       48   47   52000       14            1.6   1    0  1  0       0   
48       49   25   31000       42            3.4   0    1  0  1       1   
49       50   20   24000       33            4.7   1    0  0  1       1   

    Comedy  Drama  
0        0      0  
1        0      1  
2        1      0  
3        0      1  
4        0      0  
5        0      0  
6        0      1  
7        0      0 

In [28]:
spreadsheet.to_csv(".\VD_spreadsheet.csv")

## Crearing Co- Relation Matrix.


In [29]:
# removing Customer ID coloum
del spreadsheet['Cust_ID']

In [30]:
import matplotlib.pyplot as plt

In [31]:
spreadsheet.corr("pearson").round(2)

Age  Income  Rentals  Avg_Per_Visit    No   Yes     F     M  \
Age            1.00    0.63    -0.55           0.40  0.21 -0.21  0.23 -0.23   
Income         0.63    1.00    -0.27           0.31  0.06 -0.06  0.06 -0.06   
Rentals       -0.55   -0.27     1.00           0.07 -0.30  0.30  0.03 -0.03   
Avg_Per_Visit  0.40    0.31     0.07           1.00 -0.08  0.08  0.03 -0.03   
No             0.21    0.06    -0.30          -0.08  1.00 -1.00  0.20 -0.20   
Yes           -0.21   -0.06     0.30           0.08 -1.00  1.00 -0.20  0.20   
F              0.23    0.06     0.03           0.03  0.20 -0.20  1.00 -1.00   
M             -0.23   -0.06    -0.03          -0.03 -0.20  0.20 -1.00  1.00   
Action        -0.48   -0.36     0.33          -0.10 -0.39  0.39 -0.30  0.30   
Comedy         0.02    0.07    -0.03          -0.07  0.40 -0.40  0.02 -0.02   
Drama          0.46    0.29    -0.29           0.16  0.03 -0.03  0.28 -0.28   

               Action  Comedy  Drama  
Age             -0.48    0.02   0.46  
Income          -0.36    0.07   0.29  
Rentals          0.33   -0.03  -0.29  
Avg_Per_Visit   -0.10   -0.07   0.16  
No              -0.39    0.40   0.03  
Yes              0.39   -0.40  -0.03  
F               -0.30    0.02   0.28  
M                0.30   -0.02  -0.28  
Action           1.00   -0.42  -0.61  
Comedy          -0.42    1.00  -0.46  
Drama           -0.61   -0.46   1.00

## Co Relation Analaysis.

    The  positive numbers betweent the arrtibutes are considered to be having a Positively co related where as the Negative numbers between the attributes are considered as Negatively Co related.
    The Age and Income attribute are the most co-related arrtibute with 0.63 the followed by the drama and age attribute with 0.46.
    The least co related attribute are action and drama with -0.61 drama and comedy with -0.46